packages/@eventual/core-runtime/src/transaction-executor.ts

```diff
@@ -179,10 +174,46 @@ export function createTransactionExecutor(
                   normalizedKey
                 );
 
+                /**
+                 * When a set or delete is performed with an explicit expected version, immediately validate that the
+                 * entity resolved has that expected version.
+                 *
+                 * If a set or delete has already been performed, we'll replace that operation with this one
+                 * so we always use the original version.
+                 */
+                const expectedVersion = eventual.params[1]?.expectedVersion;
+                if (
+                  eventual.params[1]?.expectedVersion &&
+                  entityValue.originalVersion !==
+                    eventual.params[1]?.expectedVersion
+                ) {
+                  throw new UnexpectedVersion(
+                    `Operation expected version ${expectedVersion} but found ${entityValue.originalVersion}.`
+                  );
+                }
+
                 entityCalls.set(serializedKey, eventual);
-                return isEntityOperationOfType("set", eventual)
-                  ? { version: (entityValue?.version ?? 0) + 1 }
-                  : undefined;
+                if (isEntityOperationOfType("set", eventual)) {
+                  const newVersion = entityValue.originalVersion + 1;
+                  retrievedEntities.set(serializedKey, {
+                    entityName: eventual.entityName,
+                    key: normalizedKey,
+                    currentValue: eventual.params[0],
+                    currentVersion: newVersion,
+                    originalVersion: entityValue.originalVersion,
+                  });
+                  return { version: newVersion };
```

Comment:
we don't return the item on `set`?

Reply:
We don't. Which is good for transactions. What value is there to pass the input back out?

Reply:
I can think of some for update but not set